src/VisualStudio/Core/Def/Implementation/ProjectSystem/MetadataReferences/VisualStudioMetadataReferenceManager.MetadataCache.cs

```diff
@@ -48,22 +51,35 @@ private bool TryGetMetadata_NoLock(FileKey key, out AssemblyMetadata metadata)
                 return false;
             }
 
-            public bool TryGetOrAddMetadata(FileKey key, ValueSource<AssemblyMetadata> newMetadata, out AssemblyMetadata metadata)
+            /// <summary>
+            /// Gets specified metadata from the cache, or retrieves metadata from given <paramref name="metadataSource"/>
+            /// and adds it to the cache if it's not there yet.
+            /// 
+            /// <paramref name="metadataSource"/> is epxected to to provide metadata at least until this method returns.
+            /// </summary>
+            /// <returns>
+            /// True if the metadata is retrieved from <paramref name="metadataSource"/> source, false if it already exists in the cache.
+            /// </returns>
+            public bool GetOrAddMetadata(FileKey key, OptionalValueSource<AssemblyMetadata> metadataSource, out AssemblyMetadata metadata)
```

Comment:
💭 There appears to be a mismatch between the signature and the preconditions for this method. I need to think about this one some more.

Reply:
Do you mean this?
```
// the source is expected to keep the metadata alive at this point
Contract.ThrowIfNull(newMetadata);
```